crates/symbolicator/src/utils/futures.rs

```diff
@@ -235,6 +235,102 @@ where
     }
 }
 
+/// A guard to [`measure`] the amount of time it takes to download a source. This guard is also
+/// capable of calculating and reporting the the throughput of the connection. Two metrics are
+/// emitted if `bytes_transferred` is set:
+///
+/// 1. Amount of time taken to complete the measurement
+/// 2. Connection thoroughput (bytes transferred / time taken to complete)
+///
+/// If `bytes_transferred` is not set, then only 1 is recorded.
+pub struct MeasureSourceDownload<'a> {
+    state: MeasureState,
+    task_name: &'a str,
+    source_name: &'a str,
+    creation_time: Instant,
+    bytes_transferred: Option<u64>,
+}
+
+impl<'a> MeasureSourceDownload<'a> {
+    /// Creates a new measure guard for downloading a source.
+    pub fn new(task_name: &'a str, source_name: &'a str) -> Self {
+        Self {
+            state: MeasureState::Pending,
+            task_name,
+            source_name,
+            bytes_transferred: None,
+            creation_time: Instant::now(),
+        }
+    }
+
+    /// A checked add to the amount of bytes transferred during the download.
+    ///
+    /// This value will be emitted when the download's future is completed or cancelled.
+    pub fn add_bytes_transferred(&mut self, additional_bytes: u64) {
+        self.bytes_transferred = self
+            .bytes_transferred
+            .and_then(|old_count| old_count.checked_add(additional_bytes).or(Some(old_count)))
+            .or(Some(additional_bytes));
+    }
+
+    /// Marks the download as terminated.
+    pub fn done(mut self, status: &'static str) {
+        self.state = MeasureState::Done(status);
+    }
+}
+
+impl Drop for MeasureSourceDownload<'_> {
+    fn drop(&mut self) {
+        let status = match self.state {
+            MeasureState::Pending => "canceled",
+            MeasureState::Done(status) => status,
+        };
+
+        let duration = self.creation_time.elapsed().as_secs();
+        metric!(
+            histogram(self.task_name) = duration,
```

Comment:
Spoke to @flub about this on the side, and it turns out I misunderstood things here. It looks like timers (`metric!(timer(...))`) under-the-hood are simply histograms, so it makes more sense to just use that here, instead of shoving the duration into a histogram. The diff has been updated to do exactly that.